extras/create_snapshots_reports_scorecard.py

```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
+    return snapshot_process.returncode
+
+
+def create_snapshots(org_list, db, cyhy_db_section):
+    """Create a snapshot for each organization in a list."""
+    for org_id in org_list:
+        logging.info(
+            "[%s] Starting snapshot for: %s", threading.current_thread().name, org_id
         )
-        data, err = output.communicate("yes")
-        snap_time = time.time() - snap_time
-        longest_snaps.append((i, snap_time))
-        return_code = output.returncode
+        create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots=False)
 
-        if return_code == 0:
-            successful_snaps.append(i)
-            # TODO: If the org has children & their requestdoc has CYHY, add to descendant org list
-            successful_descendant_snaps += db.RequestDoc.get_all_descendants(i)
-            logging.info(
-                "Added to successful snaps: %s (%.2f s)", i, round(snap_time, 2)
+
+def generate_weekly_snapshots(db, cyhy_db_section):
+    """Generate all snapshots needed in order to generate the CyHy reports."""
+    start_time = time.time()
+
+    logging.info("Building list of reports to generate...")
+    reports_to_generate = create_list_of_reports_to_generate(db)
+
+    logging.info("Building list of snapshots to generate...")
+    snapshots_to_generate = create_list_of_snapshots_to_generate(
+        db, reports_to_generate
+    )
+
+    # List to keep track of our snapshot creation threads
+    snapshot_threads = list()
+
+    # Lists of orgs for each snapshot thread to process
+    snapshots_to_generate = list(
+        make_list_chunks(snapshots_to_generate, SNAPSHOT_THREADS)
+    )
+
+    # Start up the threads to create snapshots
+    for orgs in snapshots_to_generate:
+        try:
+            snapshot_thread = threading.Thread(
+                target=create_snapshots, args=(orgs, db, cyhy_db_section),
             )
-        else:
-            failure_snaps.append(i)
-            logging.info("Added to failed snaps: %s", i)
-            logging.info("Stderr failure detail: %s%s", data, err)
+            snapshot_threads.append(snapshot_thread)
+            snapshot_thread.start()
+        except:
+            print("Error: Unable to start snapshot thread for %s", orgs)
 
-    longest_snaps.sort(key=lambda tup: tup[1], reverse=True)
+    # Wait until each thread terminates
+    for snapshot_thread in snapshot_threads:
+        snapshot_thread.join()
+
+    snapshot_durations.sort(key=lambda tup: tup[1], reverse=True)
     logging.info("Longest Snapshots:")
-    for i in longest_snaps[:10]:
-        logging.info("%s: %s seconds", i[0], str(round(i[1], 1)))
+    for i in snapshot_durations[:10]:
+        logging.info("%s: %.1f seconds", i[0], i[1])
     logging.info(
-        "Time to complete snapshots: %.2f minutes", (round(time.time() - start, 1) / 60)
+        "Time to complete snapshots: %.2f minutes", (time.time() - start_time) / 60
```

Comment:
Just as a preference thing, but I think end times should be when work is done and not at the tail of outputting information about the work done. Thoughts?

Reply:
I could go either way on this one:

- Since the output is labeled "Time to complete _snapshots_" then it's more accurate to have the stop time be at the end of processing instead of processing + printing
- On the other hand, the process of printing out the prior information should be fairly negligible and this saves an extra variable by being able to use `time.time()` directly (again, the impact of which is negligible on our modern hardware)

Reply:
I'm in agreement with both of you, but it's easy enough to fix this, so I went ahead and did so in https://github.com/cisagov/cyhy-reports/pull/61/commits/01e36ead3e9b0e21efbf37fc2f8583d0d366ea47.  Thanks for the feedback!